### ISSUE TITLE
Subtrait blanket implementation error inferring associated type bounds

### ISSUE THREAD

**Issue by jflatow** [rust-lang/rust#116720]
This might be known behavior but I found the result rather unexpected and hard to track down. So if this is defined behavior and not a bug, at least this report might help someone figure that out.

It might be a weird pattern to define a blanket implementation of a supertrait for a subtrait, but it does seem to normally work. One benefit of doing so is that one does not need to cast when using types of the subtrait (see `W` below). However, if you do so with a certain setup of associated type bounds, you get into a pretty odd state.

I tried this code:

```rust
use std::ops::Mul;

pub trait M<P>: Mul<P, Output = Self> {}

pub trait J: M<Self::P> {
    type P;
}

trait S {
    type P;
    type V: M<Self::P>;
}

trait T: S { // only happens if T: S
    type J: J;
}

impl<I: J, U: T<J = I>> S for U { // requires I: J + M<<U as S>::P>
    type P = I::P;
    type V = I;
}

trait W<U: T> {
    fn j() -> U::P; // if T: S, otherwise <U as S>::P
}
```

I expected to this to compile. I can get it to do so by dropping the `T: S` (and casting in W appropriately), or by adding an additional bound to I in the impl, but it seems redundant and like it should be unnecessary.

Instead, the compiler fails to recognize that `I` must be `M<<U as S>::P>`. As the compiler is able to know this under other circumstances (namely without explicit `T: S`), this seems like a bug due to the circularity and errant behavior.

### Meta

This happens on both `+nightly` and `+stable`.

```
$ rustc +stable --version --verbose
rustc 1.72.1 (d5c2e9c34 2023-09-13)
binary: rustc
commit-hash: d5c2e9c342b358556da91d61ed4133f6f50fc0c3
commit-date: 2023-09-13
host: x86_64-apple-darwin
release: 1.72.1
LLVM version: 16.0.5

$ rustc +nightly --version --verbose
rustc 1.75.0-nightly (187b8131d 2023-10-03)
binary: rustc
commit-hash: 187b8131d4f760f856b214fce34534903276f2ef
commit-date: 2023-10-03
host: x86_64-apple-darwin
release: 1.75.0-nightly
LLVM version: 17.0.2
```

<details><summary>Backtrace</summary>
<p>

```
error[E0277]: the trait bound `I: M<<U as S>::P>` is not satisfied
  --> assoc-bounds.rs:20:14
   |
20 |     type V = I;
   |              ^ the trait `M<<U as S>::P>` is not implemented for `I`
   |
note: required by a bound in `S::V`
  --> assoc-bounds2.rs:11:13
   |
11 |     type V: M<Self::P>;
   |             ^^^^^^^^^^ required by this bound in `S::V`
help: consider further restricting this bound
   |
18 | impl<I: J + M<<U as S>::P>, U: T<J = I>> S for U { // requires I: J + M<<U as S>::P>
   |           ++++++++++++++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

</p>
</details>

**Comment by fmease** [comment]
Fixed by the next-gen trait solver (`-Znext-solver`).
